scripts/deploy-federation.sh

```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env bash
+
+# This script automates deployment of a federation - as documented
+# in the README - to the current kubectl context.
+#
+# If using minikube, a cluster must be started prior to invoking this
+# script:
+#
+#   $ minikube start
+#
+# This script depends on crinit, kubectl, and apiserver-boot being
+# installed in the path.  These binaries can be installed via the
+# download-test-binaries.sh script, which downloads to ./bin:
+#
+#   $ ./scripts/download-test-binaries.sh
+#   $ export PATH=$(pwd)/bin:${PATH}
+#
+# To redeploy federation from scratch, prefix the deploy invocation with the deletion script:
+#
+#   # WARNING: The deletion script will remove federation and cluster registry data
+#   $ ./scripts/delete-federation.sh && ./scripts/deploy-test-federation.sh <image>
+#
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+source "$(dirname "${BASH_SOURCE}")/util.sh"
+
+NS=federation
+IMAGE_NAME="${1:-}"
+
+if [[ -z "${IMAGE_NAME}" ]]; then
+  >&2 echo "Usage: $0 <image>
+
+<image> should be in the form <containerregistry>/<username>/<imagename>:<tagname>
+
+Example: docker.io/<username>/federation-v2:test
+
+If intending to use the docker hub as the container registry to push
+the federation image to, make sure to login to the local docker daemon
+to ensure credentials are available for push:
+
+  $ docker login --username <username>
+"
+  exit 2
+fi
+
+# Wait for the storage provisioner to be ready.  If crinit is called
+# before dynamic provisioning is enabled, the pvc for etcd will never
+# be bound.
+function storage-provisioner-ready() {
+  result="$(kubectl -n kube-system get pod storage-provisioner -o jsonpath='{.status.conditions[?(@.type == "Ready")].status}' 2> /dev/null)"
+  [[ "${result}" = "True" ]]
+}
+util::wait-for-condition "storage provisioner readiness" 'storage-provisioner-ready' 60
+
+crinit aggregated init mycr
+
+kubectl create ns "${NS}"
+
+apiserver-boot run in-cluster --name "${NS}" --namespace "${NS}" --image "${IMAGE_NAME}" --controller-args="-logtostderr,-v=4"
+
+# Create a permissive rolebinding to allow federation controllers to run.
+# TODO(marun) Make this more restrictive.
+kubectl create clusterrolebinding federation-admin --clusterrole=cluster-admin --serviceaccount="${NS}:default"
```

Comment:
For more restrictive access, we need just the following for the API server:

```
kubectl create rolebinding -n kube-system \
    federation.k8s.io:extension-apiserver-authentication-reader \
    --role=extension-apiserver-authentication-reader \
    --serviceaccount=federation:default

kubectl create clusterrolebinding federation.k8s.io:apiserver-auth-delegator \
    --clusterrole=system:auth-delegator \
    --serviceaccount=federation:default
```

For the controller manager, we need a new cluster role that includes all the verbs for the operations we want the controller manager to do on the `federation.k8s.io` resources. Then, a new cluster role binding from the service account `federation:default` to this new cluster role we just created. I don't have the exact commands for this yet.

Anyway, this one permissive cluster role binding works easily enough for now.

Reply:
I would avoid granting these cluster-admin to the default service account - recommend making a federation-controller service account and granting these perms to that. Otherwise any pod in this NS will run with a highly privileged service account token.

Reply:
I'm going to leave this as a TODO for now, with an explicit warning in the header.  Nobody should be deploying this to a production cluster in its current state.